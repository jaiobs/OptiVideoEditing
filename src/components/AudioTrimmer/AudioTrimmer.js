import {
  Dimensions,
  Image,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import React, {Component} from 'react';

import {Colors, Images} from '../../res';
import Modal from 'react-native-modal';
import TrimmerSlider from './TrimmerSlider';
import {throttle} from 'lodash/fp';

const {width: screenWidth, height: screenHeight} = Dimensions.get('window');

const maxTrimDuration = 60000;
const minimumTrimDuration = 5000;

export default class AudioTrimmer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      playing: false,
      startPosition: 0,
      startTime: 0.0,
      startedRecording: false,
      totalDuration: 140995,
      trimmerLengthOptionIndex: 1,
      startPressed: false,
      exitPressed: false,
      initialPlay: false,
    };
    this.onTrimmerValueChangedThrottles = throttle(500)(
      this.onTrimmerValueChanged,
    );
  }

  togglePlayButton = () => {
    const {playing} = this.state;

    if (playing) {
      this.trimmerRef.stopTrackProgressAnimation();
      this.setState({playing: false});
    } else {
      this.trimmerRef.startTrackProgressAnimation();
      this.setState({playing: true});
    }
  };

  onTrimmerValueChanged = (value) => {
    this.setState({startPosition: value});
  };

  changeTrimmerLength = () => {
    this.trimmerRef.stopTrackProgressAnimation();
    this.setState(
      {
        trimmerLengthOptionIndex:
          (this.state.trimmerLengthOptionIndex + 1) % TRIMMER_LENGTHS.length,
      },
      () => {
        if (this.state.playing) {
          this.trimmerRef.startTrackProgressAnimation();
        }
      },
    );
  };

  onSlidingComplete = (value) => {
    this.setState({scrubbing: false});

    if (this.trimmerRef && this.trimmerRef.scrollViewRef) {
      const {totalDuration, trimmerLengthOptionIndex} = this.state;

      const newStartingPosition =
        value *
        (totalDuration - TRIMMER_LENGTHS[trimmerLengthOptionIndex].value);

      this.setState({startPosition: newStartingPosition});
    }
    if (this.state.playing) {
      this.trimmerRef.startTrackProgressAnimation();
    }
  };

  onScrollBeginDrag = () => {
    this.trimmerRef.stopTrackProgressAnimation();
  };

  componentDidMount() {
    this.setState({
      startedRecording: false,
    });
  }

  componentWillUnMount() {
    clearInterval(this.trimmerRef.scrubberInterval);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.visibility == true && this.state.initialPlay == false) {
      this.setState(
        {
          initialPlay: true,
          startTime: 0.0,
        },
        () => {
          this.trimmerRef.startTrackProgressAnimation();
          this.props.getTrimmedStartValue(0);
        },
      );
    } else if (nextProps.visibility == false && this.state.initialPlay) {
      this.setState({
        initialPlay: false,
      });
    }
  }

  render() {
    const {
      trimmerLeftHandlePosition,
      trimmerRightHandlePosition,
      scrubberPosition,
    } = this.state;
    return (
      <Modal
        backdropColor={'transparent'}
        style={{justifyContent: 'flex-end', margin: 0, zIndex: 999}}
        isVisible={this.props.visibility}
        onModalWillHide={() => {
          this.props.onModalWillHide(
            this.state.startPressed,
            this.state.exitPressed,
          );
        }}
        onBackdropPress={this.props.onBackdropPress}>
        <View
          style={
            this.props.orientation == 'portrait'
              ? styles.containerPortrait
              : styles.containerLandscape
          }>
          {/* TITLE TRACK */}

          <View
            style={{
              flexDirection: 'row',
              paddingTop: 20,
              paddingBottom: 10,
              flex: 1,
              justifyContent: 'center',
              alignSelf: 'center',
            }}>
            <Image
              style={{height: 50, width: 50, resizeMode: 'contain'}}
              source={
                this.props.dataItem.artwork_url != null
                  ? {uri: this.props.dataItem.artwork_url}
                  : Images.song_placeholder
              }
            />

            <Text style={styles.trackTitle}>{this.props.dataItem.title}</Text>
          </View>

          {/* STARTING FROM LABEL */}
          <Text style={[styles.trackTitle, {fontSize: 12, marginLeft: 20}]}>
            start from {this.state.startTime}
          </Text>

          {/* AUDIO WAVEFORM AND TRIMMER */}
          <TrimmerSlider
            ref={(ref) => (this.trimmerRef = ref)}
            onHandleChange={this.onHandleChange}
            totalDuration={this.props.dataItem.duration}
            trimmerLeftHandlePosition={trimmerLeftHandlePosition}
            trimmerRightHandlePosition={trimmerRightHandlePosition}
            minimumTrimDuration={minimumTrimDuration}
            maxTrimDuration={maxTrimDuration}
            maximumZoomLevel={0}
            zoomMultiplier={20}
            initialZoomValue={2}
            scaleInOnInit={false}
            trimmerLength={this.props.dataItem.duration}
            orientation={this.props.orientation}
            tintColor="#f638dc"
            markerColor="#5a3d5c"
            dataItem={this.props.dataItem}
            trackBackgroundColor="#382039"
            trackBorderColor="#5a3d5c"
            scrubberColor={Colors.primaryAccent}
            scrubberPosition={scrubberPosition}
            onScrubbingComplete={this.onScrubbingComplete}
            onLeftHandlePressIn={() => console.log('onLeftHandlePressIn')}
            onRightHandlePressIn={() => console.log('onRightHandlePressIn')}
            onScrubberPressIn={() => console.log('onScrubberPressIn')}
            onScroll={(xOffset) => {
              let percentage = this.props.orientation == 'portrait'?(xOffset / (screenWidth * 2 ))*100 : (xOffset / (screenHeight * 2 ))*100
              let factorValue =((this.props.dataItem.duration / 1000)/100)*percentage
            this.setState(
              {
                startTime:
                  factorValue / 60 <= 0 ? 0.0 : (factorValue / 60).toFixed(2),
              },
              () => {
                this.props.getTrimmedStartValue(factorValue);
              },
            );
            }}
          />
          {/* FOOTER BUTTONS */}
          <View
            style={{
              flex: 1,
              alignSelf: 'center',
              flexDirection: 'row',
              marginTop: this.props.orientation == 'portrait' ? 40 : 10,
            }}>
            <TouchableOpacity
              style={styles.sideBtns}
              onPress={() => {
                this.setState(
                  {
                    startPressed: true,
                    exitPressed: false,
                  },
                  () => {
                    this.props.onBackdropPress();
                    this.trimmerRef.stopTrackProgressAnimation();
                  },
                );
              }}>
              <Text style={[styles.tabTitle, {fontSize: 16}]}>start</Text>
            </TouchableOpacity>
            <TouchableOpacity
              style={[styles.sideBtns, {marginLeft: 30}]}
              onPress={() => {
                this.setState(
                  {
                    startPressed: false,
                    exitPressed: true,
                  },
                  () => {
                    this.props.onBackdropPress();
                    this.trimmerRef.stopTrackProgressAnimation();
                  },
                );
              }}>
              <Text style={[styles.tabTitle, {fontSize: 16}]}>exit</Text>
            </TouchableOpacity>
          </View>
        </View>
      </Modal>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 0,
    alignItems: 'center',
    justifyContent: 'flex-end',
    flexDirection: 'column',
    height: '100%',
    width: '100%',
  },
  containerPortrait: {
    backgroundColor: 'rgba(0, 0, 0, 0.82)',
    height: '50%',
    width: '100%',
    alignSelf: 'center',
    zIndex: 999,
    borderTopLeftRadius: 30,
    borderTopRightRadius: 30,
  },
  containerLandscape: {
    backgroundColor: 'rgba(0, 0, 0, 0.82)',
    height: '70%',
    width: '100%',
    zIndex: 999,
    alignSelf: 'center',
    borderTopLeftRadius: 30,
    borderTopRightRadius: 30,
  },
  headerContainer: {
    backgroundColor: Colors.black,
  },
  sideBtns: {
    width: 80,
    backgroundColor: Colors.primaryAccent,
    height: 35,
    borderRadius: 20,
    justifyContent: 'center',
  },
  trackTitle: {
    color: 'white',
    marginLeft: 5,
    width: '70%',
    textAlign: 'left',
    fontSize: 14,
  },
  tabTitle: {
    color: Colors.white,
    fontSize: 14,
    marginTop: 4,
    alignSelf: 'center',
  },
  timesContainer: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  timesInnerContainer: {
    width: 150,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  timesLabel: {
    fontSize: 12,
    color: '#666',
    letterSpacing: 1,
  },
});
