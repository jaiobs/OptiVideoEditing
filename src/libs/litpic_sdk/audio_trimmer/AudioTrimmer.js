import React, {Component} from 'react';
import {
  View,
  requireNativeComponent,
  findNodeHandle,
  DeviceEventEmitter,
  UIManager,
  StyleSheet,
} from 'react-native';

const AudioTrimmerPreview = requireNativeComponent('AudioTrimmerDroid');

export default class AudioTrimmer extends Component {
  constructor(props) {
    super(props);
  }

  UNSAFE_componentWillMount() {
    this.initListeners();
  }

  initListeners() {
    this.audioStartPressListener = DeviceEventEmitter.addListener('onAudStartPressed', (audioData) => {
      this.props.onStartPressed(audioData);
    });

    this.onAudExitPressListener = DeviceEventEmitter.addListener('onAudExitPressed', (audioData) => {
      this.props.onExitPressed(audioData);
    });

    this.onAudioTrimmingListener = DeviceEventEmitter.addListener('onAudioTrimmingCompleted', (audioData) => {
      this.props.onTrimmingCompleted(audioData);
    });

    this.loaderVisibilityListener = DeviceEventEmitter.addListener(
      'EventShowOrHideLoader',
      (loaderMap) => {
        this.props.showOrHideLoader(loaderMap.showLoader);
      },
    );
  }

  componentWillUnmount() {
    this.audioStartPressListener.remove();
    this.onAudExitPressListener.remove();
    this.loaderVisibilityListener.remove();
    this.releaseNativeListeners();
  }

  onBackPressed(){
    this.executeCommand("onBackPressed",[]);
  }

  releaseNativeListeners(){
    this.executeCommand("releaseListeners",[]);
  }

  executeCommand(command, params) {
    UIManager.dispatchViewManagerCommand(
      this._audioTrimmerHandle,
      command,
      params,
    );
  }

  _setReference = (ref: ?Object) => {
    if (ref) {
      this._audioTrimmerRef = ref;
      this._audioTrimmerHandle = findNodeHandle(ref);
    } else {
      this._audioTrimmerRef = null;
      this._audioTrimmerHandle = null;
    }
  };

  render() {
    return (
      <View style={styles.audio_trimmer}>
        <AudioTrimmerPreview ref={this._setReference} {...this.props} />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  audio_trimmer: {
    position: 'absolute',
    flex: 1,
    bottom: 0,
    left: 4,
    right: 4,
    zIndex: 10,
  },
});
