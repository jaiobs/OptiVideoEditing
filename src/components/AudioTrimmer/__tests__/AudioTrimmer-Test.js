import React from 'react';
import renderer from 'react-test-renderer';
import AudioTrimmer from '../AudioTrimmer';
import TrimmerSlider from '../TrimmerSlider';

jest.useFakeTimers();
describe("AudioTrimmer Tests", () => {
    it("should render AudioTrimmer correctly", () => {
        const tree = renderer
            .create(
                <AudioTrimmer
                    dataItem={
                        { waveform_url: "" }
                    }
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    })
    it("test scrubberPosition Value", () => {
        const tree = renderer
            .create(
                <TrimmerSlider  dataItem={
                    { waveform_url: "" }
                }/>,
            )
            .getInstance();
        expect(tree.state.scrubberPosition).toEqual(1)
        tree.startTrackProgressAnimation();
        expect(tree.state.scrubberPosition).toBeDefined()
        tree.stopTrackProgressAnimation();
    })

    it("should render TrimmerSlider correctly", () => {
        const tree = renderer
            .create(
                <TrimmerSlider
                dataItem={
                    { waveform_url: "" }
                }
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    })
})