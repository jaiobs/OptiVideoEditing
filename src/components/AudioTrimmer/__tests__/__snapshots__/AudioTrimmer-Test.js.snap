// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioTrimmer Tests should render AudioTrimmer correctly 1`] = `
<Modal
  animationType="none"
  deviceHeight={null}
  deviceWidth={null}
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={false}
  onBackdropPress={[Function]}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  swipeThreshold={100}
  transparent={true}
  visible={false}
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "transparent",
        "bottom": 0,
        "height": 1334,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 750,
      }
    }
  />
  <View
    deviceHeight={null}
    deviceWidth={null}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    pointerEvents="box-none"
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    style={
      Object {
        "flex": 1,
        "justifyContent": "flex-end",
        "margin": 0,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
        "zIndex": 999,
      }
    }
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
  >
    <View
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "rgba(0, 0, 0, 0.82)",
          "borderTopLeftRadius": 30,
          "borderTopRightRadius": 30,
          "height": "70%",
          "width": "100%",
          "zIndex": 999,
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingBottom": 10,
            "paddingTop": 20,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/res/assets/images/music.png",
            }
          }
          style={
            Object {
              "height": 50,
              "resizeMode": "contain",
              "width": 50,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 14,
              "marginLeft": 5,
              "textAlign": "left",
              "width": "70%",
            }
          }
        />
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 14,
              "marginLeft": 5,
              "textAlign": "left",
              "width": "70%",
            },
            Object {
              "fontSize": 12,
              "marginLeft": 20,
            },
          ]
        }
      >
        start from 
        0
      </Text>
      <View
        style={
          Object {
            "height": 160,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": "100%",
                "justifyContent": "center",
                "paddingTop": 10,
                "position": "absolute",
                "width": 14,
                "zIndex": 1,
              },
              Object {
                "left": 0,
                "position": "absolute",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#63707e",
                  "borderRadius": 3,
                  "height": 155,
                  "marginLeft": -5,
                  "marginTop": 15,
                  "width": 5,
                },
                Object {
                  "backgroundColor": "#f9089b",
                },
              ]
            }
          />
        </View>
        <RCTScrollView
          contentContainerStyle={
            Object {
              "flexGrow": 1,
              "width": 1600,
            }
          }
          horizontal={true}
          onLayout={[Function]}
          onMoveShouldSetResponder={[Function]}
          onMoveShouldSetResponderCapture={[Function]}
          onResponderEnd={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderReject={[Function]}
          onResponderRelease={[Function]}
          onResponderStart={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onScroll={[Function]}
          onScrollEndDrag={[Function]}
          onStartShouldSetResponder={[Function]}
          onStartShouldSetResponderCapture={[Function]}
          scrollEnabled={true}
          showsHorizontalScrollIndicator={true}
          style={
            Array [
              Object {
                "height": 160,
                "marginLeft": -23,
                "overflow": "hidden",
                "position": "relative",
              },
              Object {
                "transform": Array [
                  Object {
                    "scaleX": 1,
                  },
                ],
              },
            ]
          }
        >
          <View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#f2f6f5",
                    "borderColor": "#c8dad3",
                    "borderRadius": 5,
                    "borderWidth": 1,
                    "height": 140,
                    "marginHorizontal": 31,
                    "marginVertical": 20,
                    "overflow": "hidden",
                  },
                  Object {
                    "backgroundColor": "#382039",
                    "borderColor": "#5a3d5c",
                    "width": 1500,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "height": "100%",
                    "width": "100%",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "transparent",
                    }
                  }
                >
                  <Image
                    source={
                      Object {
                        "uri": "",
                      }
                    }
                    style={
                      Object {
                        "borderLeftColor": "#f9089b",
                        "borderLeftWidth": 1,
                        "height": "100%",
                        "width": "100%",
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "flex": 1,
            "flexDirection": "row",
            "marginTop": 10,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#f9089b",
              "borderRadius": 20,
              "height": 35,
              "justifyContent": "center",
              "opacity": 1,
              "width": 80,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "color": "#fff",
                  "fontSize": 14,
                  "marginTop": 4,
                },
                Object {
                  "fontSize": 16,
                },
              ]
            }
          >
            start
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#f9089b",
              "borderRadius": 20,
              "height": 35,
              "justifyContent": "center",
              "marginLeft": 30,
              "opacity": 1,
              "width": 80,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "color": "#fff",
                  "fontSize": 14,
                  "marginTop": 4,
                },
                Object {
                  "fontSize": 16,
                },
              ]
            }
          >
            exit
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;

exports[`AudioTrimmer Tests should render TrimmerSlider correctly 1`] = `
<View
  style={
    Object {
      "height": 160,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
          "paddingTop": 10,
          "position": "absolute",
          "width": 14,
          "zIndex": 1,
        },
        Object {
          "left": 0,
          "position": "absolute",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#63707e",
            "borderRadius": 3,
            "height": 155,
            "marginLeft": -5,
            "marginTop": 15,
            "width": 5,
          },
          Object {
            "backgroundColor": "#63707e",
          },
        ]
      }
    />
  </View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "width": 1600,
      }
    }
    horizontal={true}
    onLayout={[Function]}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onScroll={[Function]}
    onScrollEndDrag={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    scrollEnabled={true}
    showsHorizontalScrollIndicator={true}
    style={
      Array [
        Object {
          "height": 160,
          "marginLeft": -23,
          "overflow": "hidden",
          "position": "relative",
        },
        Object {
          "transform": Array [
            Object {
              "scaleX": 1,
            },
          ],
        },
      ]
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#f2f6f5",
              "borderColor": "#c8dad3",
              "borderRadius": 5,
              "borderWidth": 1,
              "height": 140,
              "marginHorizontal": 31,
              "marginVertical": 20,
              "overflow": "hidden",
            },
            Object {
              "backgroundColor": "#f2f6f5",
              "borderColor": "#c8dad3",
              "width": 1500,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "height": "100%",
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "transparent",
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "",
                }
              }
              style={
                Object {
                  "borderLeftColor": undefined,
                  "borderLeftWidth": 1,
                  "height": "100%",
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
